File: spiral_traverse/spiral_traverse.py
# O(n) time O(n) space
def spiral_traverse(array):
  output = []
  start_col = 0
  start_row = 0
  end_col = len(array[0]) - 1
  end_row = len(array) - 1

  while start_col <= end_col and start_row <= end_row:
    for col in range(start_col, end_col + 1):
      output.append(array[start_row][col])
    for row in range(start_row + 1, end_row + 1):
      output.append(array[row][end_col])
    for col in reversed(range(start_col, end_col)):
      if start_row == end_row:
        break
      output.append(array[end_row][col])
    for row in reversed(range(start_row + 1, end_row)):
      if start_col == end_col:
        break
      output.append(array[row][start_col])
    start_col += 1
    end_col -= 1
    start_row += 1
    end_row -= 1

  return output

# expect [1, 2, 3, 4, 5, 6, 7, 8, 9]
print(spiral_traverse([
  [1, 2, 3],
  [8, 9, 4],
  [7, 6, 5]
]))

File: spiral_traverse/spiral_traverse.ts
function spiralTraverse(array: number[][]): number[] {
  const output: number[] = [];
  let startCol = 0;
  let startRow = 0;
  let endCol = array[0].length - 1;
  let endRow = array.length - 1;

  while (startCol <= endCol && startRow <= endRow) {
    for (let col = startCol; col <= endCol; col++) {
      output.push(array[startRow][col]);
    }
    for (let row = startRow + 1; row <= endRow; row++) {
      output.push(array[row][endCol]);
    }
    for (let col = endCol - 1; col >= startCol; col--) {
      if (startRow === endRow) break;
      output.push(array[endRow][col]);
    }
    for (let row = endRow - 1; row > startRow; row--) {
      if (startCol === endCol) break;
      output.push(array[row][startCol]);
    }
    startCol++;
    endCol--;
    startRow++;
    endRow--;
  }

  return output;
}

console.log(
  spiralTraverse([
    [1, 2, 3],
    [8, 9, 4],
    [7, 6, 5],
  ])
); // Output: [1, 2, 3, 4, 5, 6, 7, 8, 9]
